### ISSUE TITLE
GPU support

### ISSUE THREAD

**Issue by gkorland** [FalkorDB/FalkorDB#168]
Created by: @obkhan
Source: https://github.com/RedisGraph/RedisGraph/issues/1312
Any plans for GraphBLAS GPU support such as: https://github.com/gunrock/graphblast

?

**Comment by gkorland** [comment]
Origin comment by: @jeffreylovitz
Hi @obkhan,

We use Timothy Alden's SuiteSparse implementation of GraphBLAS. Work is in progress on adding GPU support to this implementation, but I'm not sure what the expected timeframe is!

**Comment by gkorland** [comment]
Origin comment by: @ghost
I see SuiteParse has GPU supoort at least since july 2020, version 5.6.
Should we expect that redisgraph benefits from gpu automatically?

cc @jeffreylovitz

**Comment by gkorland** [comment]
Origin comment by: @alexandergunnarson
Resurrecting this thread.

~@CatalinAdlerDF — I don't think automatic GPU support is possible. AFAIK it's a fundamentally different way of working with data, and the data has to either already be in GPU memory or else shipped to the GPU from RAM.~
_Edit: See my later comment in this issue. Automatic GPU support is possible._

There's also the consideration of how to split up the work between the GPU and CPU, because some queries are more suited to run on CPU vs. GPU. Also, in the general case, only relatively small portions of the graph will be able to fit into GPU memory. A good source of inspiration for this might be [Wukong](https://github.com/SJTU-IPADS/wukong), which is a distributed, GPU-accelerated graph DB with incredible performance as reported in papers, but unfortunately appears to still be an academic / non-production-grade project. It discusses its hybrid CPU/GPU architecture [here](https://github.com/SJTU-IPADS/wukong/blob/master/docs/papers/wukong+g-atc18.pdf).

Pinging @jeffreylovitz for thoughts.

Appreciate all of your awesome work on RedisGraph!

**Comment by gkorland** [comment]
Origin comment by: @alexandergunnarson
Actually, I talked with @DrTimothyAldenDavis briefly and he mentioned the following [here](https://github.com/DrTimothyAldenDavis/SuiteSparse/issues/116):

> Once GraphBLAS exploits the GPU, it will speed up RedisGraph with no extra effort on its part in general. It might help for Redis Graph to give GraphBLAS some optional hints though. We’ll see how it works out.

Should be good news for you @CatalinAdlerDF!

**Comment by gkorland** [comment]
Origin comment by: @ghost
Definitely yes.
Thanks for not abandoning this idea.

**Comment by gkorland** [comment]
Origin comment by: @DrTimothyAldenDavis
It's far from abandoned.  We're working hard on it.

**Comment by DrTimothyAldenDavis** [comment]
Yes it's very active. I just finished the CPU JIT in v8.0. A huge part of that code (16 Kloc) is also for the GPU JIT, where all CUDA kernels will go through a jitify process.  

We have several cuda kernels working and giving us 2x to 10x speedups.  They are still drafts however and not ready for production use. You can see them in the GraphBLAS/CUDA folder.
